segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -842,13 +847,15 @@ void updateSegmentIndexOffset(long segmentId, long indexOffset, int processedSiz
          * result of toExecute, otherwise it will be a different Future which will be completed when toExecute completes.
          */
         <T> CompletableFuture<T> throttleIfNeeded(DirectSegmentAccess segment, Supplier<CompletableFuture<T>> toExecute, int updateSize) {
+            // Apply credit-based throttling only for segments that are not system-critical.
+            long totalCredits = canThrottle.test(segment) ? config.getMaxUnindexedLength() : Long.MAX_VALUE;
             AsyncSemaphore throttler;
             synchronized (this) {
                 throttler = this.throttlers.getOrDefault(segment.getSegmentId(), null);
                 if (throttler == null) {
                     val si = segment.getInfo();
                     long initialDelta = Math.max(0, si.getLength() - IndexReader.getLastIndexedOffset(si));
-                    throttler = new AsyncSemaphore(config.getMaxUnindexedLength(), initialDelta, String.format("%s-%s", containerId, segment.getSegmentId()));
+                    throttler = new AsyncSemaphore(totalCredits, initialDelta, String.format("%s-%s", containerId, segment.getSegmentId()));
```

Comment:
Make sure that when you release the throttle, you also account for this case. 

Also let's add some unit tests too so that we're covered here.